File: BomberBug.java
import info.gridworld.actor.*;
import info.gridworld.grid.*;
import java.util.*;
import java.awt.Color;

//Sound effects
import java.io.BufferedInputStream;
import java.io.FileInputStream;
import javazoom.jl.player.Player;

public class BomberBug extends Bug{
	private int maxBombs;
	boolean placeBomb;
	private int bombRadius;
	private boolean superBomb;
	ArrayList<Bomb> bombs;
	private Player player;
	
	public BomberBug(){
		this(Color.RED);
	}
	
	public BomberBug(Color c){
		this(1, c);
	}
    public BomberBug(int maxBombs, Color c){
    	setColor(c);
    	bombs = new ArrayList<Bomb>();
    	this.maxBombs = maxBombs;
    	bombRadius = 2;
    	superBomb = false;
    }
    
    public void act(){}
    public void move(){}
    public void updateBombCount(){
    	//System.out.println("bombs before updating counter: " + bombs);
    	for(int i = 0; i<bombs.size();i++)
    		if(bombs.get(i).getGrid() == null){
    			bombs.remove(i);
    			i--;
    		}

    	//System.out.println("bombs after updating counter: " + bombs);
    }
    
	public boolean placeBomb(){
		updateBombCount();
		//System.out.println("bombs.size() == " + bombs.size() + "\nmaxBombs == " + maxBombs + "\n");
		if(bombs.size() < maxBombs){
			placeBomb = true;
			play("bomb_dropped");
			return true;
		}
		return false;
	}

    public void turn(){
        setDirection(getDirection() + Location.HALF_RIGHT);
    }

    public void move(int direction){
        Grid<Actor> gr = getGrid();if(gr == null)return;
        Location loc = getLocation();
        setDirection(direction);
        Location next = loc.getAdjacentLocation(getDirection());
        if(gr.isValid(next)){
        	Actor occupant = gr.get(next);
        	if(occupant == null || occupant instanceof Bonus || occupant instanceof Fire){
        		moveTo(next);
        		play("bug_step");
        		
        		if(placeBomb){
        			Bomb bomb = new Bomb(4, bombRadius);
        			
        			if(superBomb){
        				bomb.setIsSuperBomb(true);
        				superBomb = false;
        			}
        			
        			bomb.putSelfInGrid(gr, loc);
        			bombs.add(bomb);
        			placeBomb = false;
        			play("bomb_armed");
        		}
        		
        		if(occupant instanceof Bonus){
        			addBonus(occupant);
        		}
       	 	}else if(occupant instanceof Wall){
       	 		play("hit_wall");
       	 	}else if(occupant instanceof Brick){
       	 		play("hit_brick");
       	 	}else if(occupant instanceof KittyCat){
       	 		play("cat_love_you");
       	 	}
        }else play("hit_wall");
    }
    
    public void addBonus(Actor a){
    	Bonus bonus = (Bonus) a;
    	String type = bonus.getType();
    	if(type == null) return;
    	
    	if(type.equals("ExpandBombRadius")){
    		//play("extra_long");
    		bombRadius++;
    	}else if(type.equals("AddMoreBombs")){
    		//play("more_bombs");
    		maxBombs++;
    	}else if(type.equals("SuperBomb")){
    		superBomb = true;
    	}
    }
    
    public void removeSelfFromGrid(){
    	super.removeSelfFromGrid();
    	play("dead_bug");
    }

	public void play(String filename) {
		filename = "sound\\" + filename + ".mp3";
        try {
            FileInputStream fis     = new FileInputStream(filename);
            BufferedInputStream bis = new BufferedInputStream(fis);
            player = new Player(bis);
        }
        catch (Exception e) {
            System.out.println("Problem playing file " + filename);
            System.out.println(e);
        }

        // run in new thread to play in background
        new Thread() {
            public void run() {
                try { player.play(); }
                catch (Exception e) { System.out.println(e); }
            }
        }.start();
    }

	public String toString(){
		String output = "";
		Color c = getColor();
		
		if(c.equals(Color.RED))
			output += "Red";
		else if(c.equals(Color.ORANGE))
			output += "Orange";
		else if(c.equals(Color.YELLOW))
			output += "Yellow";
		else if(c.equals(Color.GREEN))
			output += "Green";
		else if(c.equals(Color.BLUE))
			output += "Blue";
		else if(c.equals(Color.PINK))
			output += "Pink";

		return output + " Bug";
	}
}

File: src/entities/BomberBug.js
import { Entity } from '../engine/Entity.js';
import { Location } from '../engine/Location.js';
import { Bomb } from './Bomb.js';
import { Bonus } from './Bonus.js';
import { Wall } from './Wall.js';
import { Brick } from './Brick.js';
import { Fire } from './Fire.js';
import { soundManager } from '../game/SoundManager.js';
import { BOMB_TIMER } from '../game/constants.js';

/**
 * Player-controlled character.
 */
export class BomberBug extends Entity {
  constructor({ color = 'red', playerIndex = 0 } = {}) {
    super({ color });
    this.playerIndex = playerIndex;
    this.maxBombs = 1;
    this.bombRadius = 2;
    this.hasSuperBomb = false;
    this.bombs = [];
    this._placeBombFlag = false;
    /** Callback invoked on death with (playerIndex, location). */
    this._onDeath = null;
  }

  /**
   * Move in the given direction if possible.
   * If placeBomb flag is set and move succeeds, places bomb at old location.
   */
  move(direction) {
    this.direction = direction;

    if (!this.grid) return;

    const dest = this.location.adjacentLocation(direction);

    // Check if destination is valid
    if (!this.grid.isValid(dest)) return;

    const occupant = this.grid.get(dest);

    // Check what's at destination — play wall/brick hit sounds
    if (occupant instanceof Wall) {
      soundManager.play('hit_wall');
      return;
    }
    if (occupant instanceof Brick) {
      soundManager.play('hit_brick');
      return;
    }
    if (occupant instanceof Bomb) return;
    if (occupant instanceof BomberBug) return;

    // Can move into: empty, Bonus, Fire, or other generic entities
    const oldLocation = this.location;

    // Collect bonus before moving (so it's removed before moveTo)
    let collectedBonus = null;
    if (occupant instanceof Bonus) {
      collectedBonus = occupant;
      occupant.removeSelfFromGrid();
    }

    // Move
    this.moveTo(dest);
    soundManager.play('bug_step');

    // Place bomb at old location if flag is set
    if (this._placeBombFlag) {
      const bomb = new Bomb(BOMB_TIMER, this.bombRadius);
      bomb.owner = this;
      bomb.isSuperBomb = this.hasSuperBomb;
      if (this.hasSuperBomb) {
        soundManager.play('watch_out_superbomb');
        this.hasSuperBomb = false;
      }
      bomb.putSelfInGrid(this.grid, oldLocation);
      this.bombs.push(bomb);
      this._placeBombFlag = false;
      soundManager.play('bomb_dropped');
    }

    // Apply collected bonus
    if (collectedBonus) {
      this.addBonus(collectedBonus);
      soundManager.play('bonus_gotten');
    }
  }

  /**
   * Queue a bomb to be placed on next successful move.
   * Returns true if bomb can be placed, false if at max.
   */
  placeBomb() {
    this.updateBombCount();
    if (this.bombs.length >= this.maxBombs) {
      return false;
    }
    this._placeBombFlag = true;
    return true;
  }

  /**
   * Apply a bonus power-up.
   */
  addBonus(bonus) {
    switch (bonus.type) {
      case Bonus.EXPAND_RADIUS:
        this.bombRadius++;
        break;
      case Bonus.ADD_BOMBS:
        this.maxBombs++;
        break;
      case Bonus.SUPER_BOMB:
        this.hasSuperBomb = true;
        break;
    }
  }

  /**
   * Remove bombs that are no longer in the grid from the tracking array.
   */
  updateBombCount() {
    this.bombs = this.bombs.filter(bomb => bomb.grid !== null);
  }

  /**
   * Remove bug from grid (death).
   */
  removeSelfFromGrid() {
    const loc = this.location; // capture before super clears it
    soundManager.play('dead_bug');

    // Notify listeners for death animation
    if (typeof this._onDeath === 'function') {
      this._onDeath(this.playerIndex, loc);
    }

    super.removeSelfFromGrid();
  }
}
